scripts/calc_from_args.py

```diff
@@ -0,0 +1,223 @@
+        --size_in       udf='Eluted Library Size (bp)',source='step' \
+        --amt_out       udf='Library Loading Amount (fmol)',source='step' \
+        --vol_out       udf='Library Volume to Use for Loading (ul)',source='step'
+
+    Example 4:
+
+        python {__file__} \
+        --pid		    '24-885819' \
+        --calc          'equimolar_pooling' \
+        --log           'Calculate pooling log' \
+        --vol_in        udf='Eluted Volume (ul)' \
+        --conc_in       udf='Eluted Concentration (ng/ul)' \
+        --size_in       udf='Size (bp)',recursive=True \
+        --amt_out       udf='Amount (fmol)' \
+        --vol_out       udf='Total Volume (uL)'
+
+    """
+
+    # Parse args
+    parser = ArgumentParser(description=DESC)
+    parser.add_argument("--pid", type=str, help="Lims ID for current Process")
+    parser.add_argument(
+        "--calc",
+        type=str,
+        choices=["volume_to_use", "amount", "equimolar_pooling", "summarize_pooling"],
+        help="Which function to use for calculations",
+    )
+    parser.add_argument("--log", type=str, help="Which log file slot to use")
+    # UDFs to use for calculations
+    udf_args = [
+        "vol_in",
+        "size_in",
+        "conc_in",
+        "conc_units_in",
+        "amt_out",
+        "vol_out",
+        "size_out",
+    ]
+    for udf_arg in udf_args:
+        parser.add_argument(f"--{udf_arg}", type=parse_udf_arg)
+    args = parser.parse_args()
+
+    # Set up LIMS
+    lims = Lims(BASEURI, USERNAME, PASSWORD)
+    lims.check_version()
+    process = Process(lims, id=args.pid)
+
+    # Set up logging
+    log_filename: str = (
+        "_".join(
+            [
+                SCRIPT_NAME,
+                args.calc,
+                process.id,
+                TIMESTAMP,
+                process.technician.name.replace(" ", ""),
+            ]
+        )
+        + ".log"
+    )
+
+    logging.basicConfig(
+        filename=log_filename,
+        filemode="w",
+        format="%(levelname)s: %(message)s",
+        level=logging.INFO,
+    )
+
+    # Start logging
+    logging.info(f"Script '{SCRIPT_NAME}' started at {TIMESTAMP}.")
+    logging.info(
+        f"Launched in step '{process.type.name}' ({process.id}) by {process.technician.name}."
+    )
+    args_str = "\n\t".join([f"'{arg}': {getattr(args, arg)}" for arg in vars(args)])
+    logging.info(f"Script called with arguments: \n\t{args_str}")
+
+    try:
+        function_to_use = getattr(calculation_methods, args.calc)
+        function_to_use(process, args)
+    except Exception as e:
+        # Post error to LIMS GUI
+        logging.error(str(e), exc_info=True)
+        logging.shutdown()
+        upload_file(
+            file_path=log_filename,
+            file_slot=args.log,
+            process=process,
+            lims=lims,
+            remove=True,
+        )
+        sys.stderr.write(str(e))
+        sys.exit(2)
+    else:
```

Comment:
I don't think the else matters here since you're doing `sys.exit(2)` in case of error? But maybe it's clearer?

Reply:
I like how the clause shows that we have a two clear, mutually exclusive outcomes of run-time failure and run-time success respectively.